File: php/starwars_sdk.php
<?php
declare(strict_types=1);

// StarWars SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class StarWarsSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new StarWarsUtility();
        $this->_utility = $utility;

        $config = StarWarsConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = StarWarsHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = StarWarsHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, StarWarsFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return StarWarsUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = StarWarsHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = StarWarsHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = StarWarsHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new StarWarsSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = StarWarsHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = StarWarsHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_film = null;

    // Canonical facade: $client->Film()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->film()
    // resolves here too.
    public function Film($data = null)
    {
        require_once __DIR__ . '/entity/film_entity.php';
        if ($data === null) {
            if ($this->_film === null) {
                $this->_film = new FilmEntity($this, null);
            }
            return $this->_film;
        }
        return new FilmEntity($this, $data);
    }


    private $_people_list = null;

    // Canonical facade: $client->PeopleList()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->people_list()
    // resolves here too.
    public function PeopleList($data = null)
    {
        require_once __DIR__ . '/entity/people_list_entity.php';
        if ($data === null) {
            if ($this->_people_list === null) {
                $this->_people_list = new PeopleListEntity($this, null);
            }
            return $this->_people_list;
        }
        return new PeopleListEntity($this, $data);
    }


    private $_person = null;

    // Canonical facade: $client->Person()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->person()
    // resolves here too.
    public function Person($data = null)
    {
        require_once __DIR__ . '/entity/person_entity.php';
        if ($data === null) {
            if ($this->_person === null) {
                $this->_person = new PersonEntity($this, null);
            }
            return $this->_person;
        }
        return new PersonEntity($this, $data);
    }


    private $_planet = null;

    // Canonical facade: $client->Planet()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->planet()
    // resolves here too.
    public function Planet($data = null)
    {
        require_once __DIR__ . '/entity/planet_entity.php';
        if ($data === null) {
            if ($this->_planet === null) {
                $this->_planet = new PlanetEntity($this, null);
            }
            return $this->_planet;
        }
        return new PlanetEntity($this, $data);
    }


    private $_species = null;

    // Canonical facade: $client->Species()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->species()
    // resolves here too.
    public function Species($data = null)
    {
        require_once __DIR__ . '/entity/species_entity.php';
        if ($data === null) {
            if ($this->_species === null) {
                $this->_species = new SpeciesEntity($this, null);
            }
            return $this->_species;
        }
        return new SpeciesEntity($this, $data);
    }


    private $_starship = null;

    // Canonical facade: $client->Starship()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->starship()
    // resolves here too.
    public function Starship($data = null)
    {
        require_once __DIR__ . '/entity/starship_entity.php';
        if ($data === null) {
            if ($this->_starship === null) {
                $this->_starship = new StarshipEntity($this, null);
            }
            return $this->_starship;
        }
        return new StarshipEntity($this, $data);
    }


    private $_vehicle = null;

    // Canonical facade: $client->Vehicle()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->vehicle()
    // resolves here too.
    public function Vehicle($data = null)
    {
        require_once __DIR__ . '/entity/vehicle_entity.php';
        if ($data === null) {
            if ($this->_vehicle === null) {
                $this->_vehicle = new VehicleEntity($this, null);
            }
            return $this->_vehicle;
        }
        return new VehicleEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new StarWarsSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: py/starwars_sdk.py
# StarWars SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import StarWarsUtility
from core.spec import StarWarsSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import StarWarsBaseFeature
from features import _make_feature


class StarWarsSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = StarWarsUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return StarWarsUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = StarWarsSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def film(self):
        """Idiomatic facade: client.film.list() / client.film.load({"id": ...})."""
        from entity.film_entity import FilmEntity
        cached = getattr(self, "_film", None)
        if cached is None:
            cached = FilmEntity(self, None)
            self._film = cached
        return cached

    def Film(self, data=None):
        # Deprecated: use client.film instead.
        from entity.film_entity import FilmEntity
        return FilmEntity(self, data)


    @property
    def people_list(self):
        """Idiomatic facade: client.people_list.list() / client.people_list.load({"id": ...})."""
        from entity.people_list_entity import PeopleListEntity
        cached = getattr(self, "_people_list", None)
        if cached is None:
            cached = PeopleListEntity(self, None)
            self._people_list = cached
        return cached

    def PeopleList(self, data=None):
        # Deprecated: use client.people_list instead.
        from entity.people_list_entity import PeopleListEntity
        return PeopleListEntity(self, data)


    @property
    def person(self):
        """Idiomatic facade: client.person.list() / client.person.load({"id": ...})."""
        from entity.person_entity import PersonEntity
        cached = getattr(self, "_person", None)
        if cached is None:
            cached = PersonEntity(self, None)
            self._person = cached
        return cached

    def Person(self, data=None):
        # Deprecated: use client.person instead.
        from entity.person_entity import PersonEntity
        return PersonEntity(self, data)


    @property
    def planet(self):
        """Idiomatic facade: client.planet.list() / client.planet.load({"id": ...})."""
        from entity.planet_entity import PlanetEntity
        cached = getattr(self, "_planet", None)
        if cached is None:
            cached = PlanetEntity(self, None)
            self._planet = cached
        return cached

    def Planet(self, data=None):
        # Deprecated: use client.planet instead.
        from entity.planet_entity import PlanetEntity
        return PlanetEntity(self, data)


    @property
    def species(self):
        """Idiomatic facade: client.species.list() / client.species.load({"id": ...})."""
        from entity.species_entity import SpeciesEntity
        cached = getattr(self, "_species", None)
        if cached is None:
            cached = SpeciesEntity(self, None)
            self._species = cached
        return cached

    def Species(self, data=None):
        # Deprecated: use client.species instead.
        from entity.species_entity import SpeciesEntity
        return SpeciesEntity(self, data)


    @property
    def starship(self):
        """Idiomatic facade: client.starship.list() / client.starship.load({"id": ...})."""
        from entity.starship_entity import StarshipEntity
        cached = getattr(self, "_starship", None)
        if cached is None:
            cached = StarshipEntity(self, None)
            self._starship = cached
        return cached

    def Starship(self, data=None):
        # Deprecated: use client.starship instead.
        from entity.starship_entity import StarshipEntity
        return StarshipEntity(self, data)


    @property
    def vehicle(self):
        """Idiomatic facade: client.vehicle.list() / client.vehicle.load({"id": ...})."""
        from entity.vehicle_entity import VehicleEntity
        cached = getattr(self, "_vehicle", None)
        if cached is None:
            cached = VehicleEntity(self, None)
            self._vehicle = cached
        return cached

    def Vehicle(self, data=None):
        # Deprecated: use client.vehicle instead.
        from entity.vehicle_entity import VehicleEntity
        return VehicleEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
